### PR TITLE
genesys: Identify hp usbhubs with public key

### PR DESCRIPTION
**Pull request by gportay**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

The idea behind those changes is to prepare the genesys plugin for upgrading the genuine Genesys Logic Hubs products.

These products uses `USB\VID_05E3&PID_0610` which is not set in the quirk file yet; will be in a dedicated PR, once the support is fully functional.

The support for genuine Genesys Logic Hubs is splited into several PR. This one is the first of the serie.

In short, this PR:
 - Adds add the public key to the instance-id to with precision the HP hub; M24fd and M27fd have different firmwares; public-keys are differents at least.  
 - Adds the private flag `has-public-key` which tells the plugin the device support public-key and do authentication on attach/detach. This feature is specific to HP devices. The Genesys Logic products do not use this feature.
 - Sets the flag `dual-image` in the quirk file instead of hardcoding it in the plugin; the dual bank feature is not necessarily used by the hubs (HP use it).

### CODE REVIEW

**Comment by gportay** [comment]
@adamgene, can I ask for a review?

File: plugins/genesys/README.md [hunk 1]
```diff
@@ -24,7 +24,9 @@ This plugin supports the following protocol IDs:
 
 These devices use the standard USB DeviceInstanceId values for the USB Hub, e.g.
 
-* HP Mxfd FHD Monitor: `USB\VID_03F0&PID_0610`
+* HP USB-C Controller: `USB\VID_03F0&PID_0610`
```
Comment (adamgene) [inline plugins/genesys/README.md:27]: The instance ID is not for `HP USB-C Controller`, It is generic with genesys hub.

**Review by adamgene — DISMISSED** [review]
Current changes is fine.

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 2]
```diff
@@ -612,8 +624,9 @@ static gboolean
 fu_genesys_usbhub_device_detach(FuDevice *device, FuProgress *progress, GError **error)
 {
 	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
-	if (!fu_genesys_usbhub_device_authenticate(self, error))
-		return FALSE;
+	if (fu_device_has_private_flag(device, FU_GENESYS_USBHUB_FLAG_HAS_PUBLIC_KEY))
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:627]: maybe do outer braces for this

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 3]
```diff
@@ -805,8 +818,9 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 		}
 	}
 
-	if (!fu_genesys_usbhub_device_authenticate(self, error))
-		return FALSE;
+	if (fu_device_has_private_flag(device, FU_GENESYS_USBHUB_FLAG_HAS_PUBLIC_KEY))
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:821]: {} please

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 4]
```diff
@@ -919,6 +931,26 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
...
+				    self->fw_data_total_count, /* src */
+				    sizeof(self->public_key),
+				    error))
+			return FALSE;
+		guid = fwupd_guid_hash_data(self->public_key,
+					    sizeof(self->public_key),
+					    FWUPD_GUID_FLAG_NONE);
+		fu_device_add_instance_strup(device, "PUBKEY", guid);
+		if (!fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL))
+			return FALSE;
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:951]: you can do this alongside the other calls to fu_device_build_instance_id --  if `PUBKEY` isn't defined then it's a No-Op.

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 5]
```diff
@@ -962,12 +994,31 @@ fu_genesys_usbhub_device_prepare_firmware(FuDevice *device,
...
+		const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+
+		if (g_getenv("FWUPD_GENESYS_USBHUB_VERBOSE") != NULL)
+			fu_common_dump_raw(G_LOG_DOMAIN, "Footer", buf, bufsz);
+		if (memcmp(buf + fu_firmware_get_size(firmware),
+			   self->public_key,
+			   sizeof(self->public_key)) != 0) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:1016]: NOT_SUPPORTED or SIGNATURE_INVALID perhaps?

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 4 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c:951]: Well, there is no other call `fu_device_build_instance_id()` (yet :p); I have the idea.

However, I got a segmentation fault if the instance is not defined.

I made the change below:

```
--- a/plugins/genesys/fu-genesys-usbhub-device.c
+++ b/plugins/genesys/fu-genesys-usbhub-device.c
@@ -949,10 +949,11 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
                                            sizeof(self->public_key),
                                            FWUPD_GUID_FLAG_NONE);
                fu_device_add_instance_strup(device, "PUBKEY", guid);
-               if (!fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL))
-                       return FALSE;
        }

+       if (!fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL))
+               return FALSE;
+
        /* have MStar scaler */
        if (fu_device_has_private_flag(device, FU_GENESYS_USBHUB_FLAG_HAS_MSTAR_SCALER))
                if (!fu_genesys_usbhub_device_mstar_scaler_setup(self, error))
```

Do you still want that change? and/or I fix the segmentation or ???

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 3 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c:821]: done

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 2 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c:627]: done

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 5 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c]: done

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 4 above)
Comment (gportay) [reply plugins/genesys/fu-genesys-usbhub-device.c]: done

**Comment by hughsie** [comment]
@gportay if I'm honest I'm very confused about which ones come before the others...

**Comment by gportay** [comment]
> @gportay if I'm honest I'm very confused about which ones come before the others...

Pick the other (aka #4427), I will make the change then.

Sorry for the confusion :p

**Comment by gportay** [comment]
@hughsie ready to get merged!

**Review by hughsie — APPROVED** [review]
